### PR TITLE
feat: add LoHa and LoKr network modules with architecture detection

### PR DESCRIPTION
**Pull request by kohya-ss**
This pull request adds comprehensive support for LoHa (Low-rank Hadamard Product) as an alternative parameter-efficient fine-tuning method alongside LoRA in Musubi Tuner. It introduces a new `loha.py` network module, updates documentation, and makes minor improvements to the core LoRA network code to support modularity and extensibility.

**Major features and documentation:**

*Adds support for LoHa network modules and documents their usage and limitations, while refactoring the core LoRA code to allow for easy integration of new network module types.*

---

**LoHa network module implementation:**

- Introduced a new `LoHa` network module in `loha.py`, including both training (`LoHaModule`) and inference (`LoHaInfModule`) classes, custom autograd function for the Hadamard product, weight merging utilities, and architecture auto-detection. This enables training and inference using LoHa on supported architectures, with dropout and module exclusion features.

**Documentation updates:**

- Added a new documentation file `docs/loha_lokr.md` that explains LoHa and LoKr, their usage, supported architectures, training/inference instructions, options, limitations, and acknowledgments, in both English and Japanese.

**Core LoRA network extensibility:**

- Updated `lora.py` to support passing a custom `module_class` and `module_kwargs` to the `create_network` function, enabling integration of alternative network modules like LoHa. [[1]](diffhunk://#diff-f0929933f0f0297a23385c95e599a0c21689b543be6be1b443791608c9e9fb05R330-R331) [[2]](diffhunk://#diff-f0929933f0f0297a23385c95e599a0c21689b543be6be1b443791608c9e9fb05R373-R375)
- Minor type annotation improvement in `lora.py` to support `Any` type for future extensibility.

**Code cleanup:**

- Removed commented-out legacy code for weight merging in `lora.py` for clarity and maintainability.

---

Much of the code was written by Claude Code/Opus 4.6.

### CODE REVIEW

File: src/musubi_tuner/networks/lokr.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+        if "alpha" in key:
+            modules_alpha[lora_name] = value
+        elif "lokr_w2_a" in key:
+            # low-rank mode: dim = w2_a.shape[1]
+            dim = value.shape[1]
+            modules_dim[lora_name] = dim
+        elif "lokr_w2" in key and "lokr_w2_a" not in key and "lokr_w2_b" not in key:
+            # full matrix mode: use dim=1 as placeholder
+            if lora_name not in modules_dim:
+                modules_dim[lora_name] = 1
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/networks/lokr.py]: You are right. I will update the code.

File: src/musubi_tuner/networks/convert_z_image_lora_to_comfy.py [hunk 2]
```diff
@@ -139,8 +143,141 @@ def main(args):
...
+                state_dict[f"{new_lora_name}.hada_w2_a"] = w2a_qkv
+                state_dict[f"{new_lora_name}.hada_w2_b"] = w2b_qkv
+                state_dict[f"{new_lora_name}.alpha"] = alpha * 3  # rank is 3x, so alpha * 3 keeps scale unchanged
+
+                qkv_count += 1
+
+        # LoKr QKV merge: materialize weight deltas via Kronecker product, concatenate, convert to LoRA via SVD
+        # Kronecker product structure cannot be preserved across QKV concatenation,
+        # so we convert QKV layers to LoRA format. Non-QKV layers remain as LoKr.
+        keys = list(state_dict.keys())
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/networks/convert_z_image_lora_to_comfy.py:192]: ComfyUI seems to support mixed weights, so I'll leave it as is for now.

File: docs/loha_lokr.md [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+</details>
+
+### Format conversion / フォーマット変換
+
+`convert_lora.py` is extended to also support format conversion of LoHa/LoKr weights between Musubi Tuner format and Diffusers format for ComfyUI.
+
+<details>
+<summary>日本語</summary>
+
+`convert_lora.py` は、LoRAに加えて、LoHa/LoKrの重みのフォーマット変換（Musubi Tuner形式とDiffusers形式間の変換）についてもサポートするよう、拡張されています。
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply docs/loha_lokr.md:339]: I've updated here.

File: src/musubi_tuner/utils/lora_utils.py [hunk 4]
```diff
@@ -22,6 +22,18 @@
...
+    for key in lora_sd:
+        if "lora_down" in key:
+            return "lora"
+        if "hada_w1_a" in key:
+            return "loha"
+        if "lokr_w1" in key:
+            return "lokr"
+    return "lora"  # default
+
+
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/utils/lora_utils.py:36]: ComfyUI seems to support mixed weights, so I'll leave it as is for now.

File: src/musubi_tuner/networks/loha.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+class HadaWeight(torch.autograd.Function):
+    """Efficient Hadamard product forward/backward for LoHa.
+
+    Computes ((w1a @ w1b) * (w2a @ w2b)) * scale with custom backward
+    that recomputes intermediates instead of storing them.
+    """
+
+    @staticmethod
+    def forward(ctx, w1a, w1b, w2a, w2b, scale=torch.tensor(1)):
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/networks/loha.py]: I've updated.
